modules/platform/nuxeo-preview-core/src/main/java/org/nuxeo/ecm/platform/preview/adapter/ZipPreviewer.java

```diff
@@ -0,0 +1,131 @@
+ *     Kevin Leturc <kleturc@nuxeo.com>
+ */
+package org.nuxeo.ecm.platform.preview.adapter;
+
+import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.LineIterator;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.Blobs;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.blobholder.BlobHolder;
+import org.nuxeo.ecm.core.api.blobholder.SimpleBlobHolder;
+import org.nuxeo.ecm.core.convert.api.ConversionException;
+import org.nuxeo.ecm.core.convert.api.ConversionService;
+import org.nuxeo.ecm.platform.htmlsanitizer.HtmlSanitizerService;
+import org.nuxeo.ecm.platform.mimetype.interfaces.MimetypeRegistry;
+import org.nuxeo.ecm.platform.preview.api.PreviewException;
+import org.nuxeo.ecm.platform.preview.helper.PreviewHelper;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Previewer for Zip blob.
+ * <p>
+ * It sanitizes each zip entry if needed.
+ *
+ * @since 11.5
+ */
+public class ZipPreviewer implements MimeTypePreviewer {
+
+    protected static final Set<String> HTML_MIME_TYPES = Set.of("text/html", "text/xml", "text/plain");
+
+    private static final Logger log = LogManager.getLogger(ZipPreviewer.class);
+
+    @Override
+    public List<Blob> getPreview(Blob blob, DocumentModel dm) throws PreviewException {
+        ConversionService conversionService = Framework.getService(ConversionService.class);
+        String converterName = conversionService.getConverterName("application/zip", "text/html");
+        if (converterName == null) {
+            throw new PreviewException("Unable to find zip2html converter");
+        }
+
+        try {
+            BlobHolder result = conversionService.convert(converterName, new SimpleBlobHolder(blob), null);
+            List<Blob> blobs = new ArrayList<>(result.getBlobs());
+            blobs.replaceAll(this::sanitize);
+            return blobs;
+        } catch (ConversionException e) {
+            throw new PreviewException(e.getMessage(), e);
+        }
+    }
+
+    protected Blob sanitize(Blob blob) throws PreviewException {
+        String filename = blob.getFilename();
+        if (!isSanitizable(blob)) {
+            log.debug("ZipEntryBlob: {} is not sanitizable", filename);
+            return blob;
+        }
+        try {
+            String content = blob.getString();
+            content = getHtmlSanitizerService().sanitizeString(content, null);
+            content = PreviewHelper.makeHtmlPage(content);
+            return Blobs.createBlob(content, "text/html", "UTF-8", filename);
+        } catch (IOException e) {
+            throw new PreviewException("Cannot fetch ZipEntryBlob content with filename: " + filename, e);
```

Comment:
Is that the right error message?

Reply:
Yes, IOException is only thrown by `blob.getString()`.